resource/resource.go

```diff
@@ -168,46 +161,202 @@ func (r *Resource) Out(inputDirectory string, source oc.Source, params oc.Params
+		return nil, nil, err
+	}
+
+	if len(p.SecretMaps) == 0 {
+		for _, file := range files {
+			p.SecretMaps = append(p.SecretMaps, SecretMap{Source: file})
+		}
+	}
+
+	for _, secretMap := range p.SecretMaps {
+		if secretMap.Dest == "" {
+			secretMap.Dest = secretMap.Source
+		}
+
+		sourceFile := filepath.Join(rootDir, secretMap.Source)
+
+		secret, err := createSecret(sourceFile)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		finalKeys, err := getFinalKeys(secretMap.Keys)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		err = validate(secret, finalKeys)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		err = filterAndRenameKeys(secret, finalKeys)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		finalVaultPath := filepath.Join(p.Prefix, secretMap.Dest)
+		retainExistingKeys(r.client, finalVaultPath, secret)
+
+		err = copySecretToVault(r.client, finalVaultPath, secret)
+		if err != nil {
+			return nil, nil, err
+		}
+
+	}
+	// Both `version` and `metadata` may be empty. In this case, we are returning
+	// `version` just as we do from `Check`, while `metadata` is empty.
+	ocVersion, err := r.constructVersion(s, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	metadata := oc.Metadata{{Name: "Version", Value: fmt.Sprint(ocVersion)}}
+	return ocVersion, metadata, nil
+}
+
+func listFilesUnder(rootDir string) ([]string, error) {
+	ret := []string{}
+	err := filepath.Walk(rootDir,
 		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
 			if info.IsDir() {
 				return nil
 			}
-
-			bytes, err := ioutil.ReadFile(path)
+			secretPath, err := filepath.Rel(rootDir, path)
 			if err != nil {
 				return err
 			}
-			secert := sv.NewSecret()
-			err = secert.UnmarshalJSON(bytes)
-			if err != nil {
-				return err
+			ret = append(ret, secretPath)
+			return nil
+		},
+	)
+	return ret, err
+}
+
+func createSecret(secretFile string) (*sv.Secret, error) {
+	srcFile, err := os.Open(secretFile)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading source file '%s': %s", secretFile, err)
+	}
+	jDec := json.NewDecoder(srcFile)
+	secret := sv.NewSecret()
+	err = jDec.Decode(&secret)
+	if err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+func validate(secret *sv.Secret, finalKeys map[string]string) error {
+	if len(finalKeys) == 0 {
+		return nil
+	}
+	aKeyIsMissing := false
+	missingKeys := []string{}
+	keys := make([]string, 0, len(finalKeys))
+	values := make([]string, 0, len(finalKeys))
+	var sb strings.Builder
+	for key, value := range finalKeys {
+		keys = append(keys, key)
+		values = append(values, value)
+		if !secret.Has(key) {
+			aKeyIsMissing = true
+			missingKeys = append(missingKeys, key)
+		}
+	}
+
+	illegalKeys := []string{}
+	illegalKeysFound := false
+	for _, value := range values {
+		for _, key := range keys {
+			if key == value {
+				illegalKeys = append(illegalKeys, key)
 			}
-			secretPath, err := filepath.Rel(rootDir, path)
+		}
+	}
```

Comment:
don't make using an existing key name illegal. In the most basic case, this makes _not_ renaming a key illegal. And in more complicated "circular reference" cases, we can avoid having it be a problem by just not using the source secret as our scratch space - i.e. copying the secret to a "dest" secret object and then only overwriting the values in that.

Reply:
This is working correctly as is. Covered by test:
It("should import only specified keys and write them back to the same path name, retaining original key names"...

Reply:
Ok, I take that back. There was actually another bug that was hiding this bug. I fixed the other one and this one has now surfaced.